tests/sources/test_notion.py

```diff
@@ -740,13 +759,85 @@ async def test_advanced_rules_validation(advanced_rules, expected_validation_res
     async with create_source(
         NotionDataSource, notion_secret_key="secret_key"
     ) as source:
-        with patch(
-            "connectors.sources.notion.async_iterate_paginated_api",
+        with patch.object(
+            NotionClient,
+            "async_iterate_paginated_api",
             return_value=AsyncIterator([DATABASE]),
         ), patch.object(
             NotionDataSource, "get_entities", return_value=[DATABASE, BLOCK]
         ):
             validation_result = await NotionAdvancedRulesValidator(source).validate(
                 advanced_rules
             )
-        assert validation_result == expected_validation_result
+            assert validation_result == expected_validation_result
+
+
+@pytest.mark.asyncio
+async def test_async_iterate_paginated_api():
+    async def mock_function(**kwargs):
+        return {
+            "results": [{"name": "John"}, {"name": "Alice"}],
+            "next_cursor": None,
+            "has_more": False,
+        }
+
+    with patch("notion_client.AsyncClient") as mock_client:
+        mock_client_instance = mock_client.return_value
+        mock_client_instance.__aenter__.return_value = mock_client_instance
+        mock_client_instance.some_function = AsyncMock(side_effect=mock_function)
+        async with create_source(
+            NotionDataSource, notion_secret_key="secert_key"
+        ) as source:
+            async for result in source.notion_client.async_iterate_paginated_api(
+                mock_client_instance.some_function
+            ):
+                assert "name" in result
+            mock_client_instance.some_function.assert_called_once()
+
+
+@patch("connectors.utils.time_to_sleep_between_retries", Mock(return_value=0))
+@pytest.mark.asyncio
+async def test_fetch_results_rate_limit_exceeded():
+    async def mock_function_with_429(**kwargs):
+        raise APIResponseError(
+            code="rate_limited",
+            message="Rate limit exceeded.",
+            response=Response(status_code=429, text="Rate limit exceeded."),
+        )
+
+    async with create_source(
+        NotionDataSource, notion_secret_key="secert_key"
+    ) as source:
+        with patch("connectors.sources.notion.DEFAULT_RETRY_SECONDS", 0.3):
+            with pytest.raises(Exception) as exc_info:
+                await source.notion_client.fetch_results(mock_function_with_429)
```

Comment:
Can you adapt this to show that it raises if it keeps getting the 429, but explicitly _does not_ raise an error if it 429s once, but then responds successfully after the retry? For example see https://github.com/elastic/connectors/blob/main/tests/sources/test_sharepoint_online.py#L571